### ISSUE TITLE
Does the ZeRO support the SGD optimizer?

### ISSUE THREAD

**Issue by guozhiyao** [microsoft/DeepSpeed#1653]
I want to speed up my face recognition model with DeepSpeed. And the classifier is MP and I want to use the DeepSpeed to the backbone. I use the SGD optimizer and ZeRO stage is 1. But the GPU memory does not reduce, while even larger with stage 2 and 3.

The following is my code form：
```
backbone = get_backbone()
classifier = get_classifier() # is MP

class_optimizer = torch.optim.SGD(classifier.parameters(), lr=0.01, momentum=0.9)
net, _, _, _ = deepspeed.initialize(args=args, model=backbone, model_parameters=net.parameters())

for batch in data_iter:
  feat = net(batch)
  loss = classifier(feat)
  class_optimizer.zero_grad()
  backbone.backward(loss)
  class_optimizer.step()
  backbone.step()
```

**Comment by tjruwase** [comment]
@guozhiyao, thanks for reporting this problem. ZeRO is most useful for large models (>1B parameters) and optimizers with large memory footprint (e.g., ADAM which maintains 16 bytes per model parameter). So, it is not surprising that you are not seeing benefits for a lightweight optimizer like SGD. How large is your model and how many GPUs are you training with?

You might find these [docs](https://deepspeed.readthedocs.io/en/latest/memory.html) useful for understanding ZeRO and memory consumption.

**Comment by tjruwase** [comment]
@guozhiyao, do you have further question on this issue or can it be closed? Thanks!

**Comment by nsriniva03** [comment]
But if ZeRO is set to 3 then it should help with gradient partitioning  and parameter partitioning. Thus enabling to train with larger batch sizes?

**Comment by tjruwase** [comment]
@nsriniva03, yes, your understanding is correct. My point was that ZeRO maintains intermediate buffers and the default size of these buffers is more reasonable relative to the high memory costs of ADAM. For SGD, the memory costs of these buffers become significant. The good news is that you can explore reducing these buffer sizes as discussed [here](https://github.com/microsoft/DeepSpeed/issues/467).
